### PR TITLE
feat: use JSONForms for complex block editing

### PR DESCRIPTION
**Pull request by dcshzj**
## Problem

Hand-rolling our own form-based editor is very difficult to do, especially when it comes to `$ref` and `oneOf`. Using JSONForms is a lot easier, and provides better type-safety.

## Solution

**Breaking Changes**

- [ ] Yes - this PR contains breaking changes
- [x] No - this PR is backwards compatible

**Features**:

- Shift the native inputs to use JSONForms instead
    - Note: There are a lot more renderers that JSONForms needs, I will be adding the more complex ones in subsequent PRs
    - The intention is to add just enough renderers for MVP. We can add the other renderers later if needed.

**References**:

- [Material](https://github.com/eclipsesource/jsonforms/tree/master/packages/material-renderers) and [vanilla renderers](https://github.com/eclipsesource/jsonforms/tree/master/packages/vanilla-renderers) of JSONForms

## Considerations

The initial plan was to use JSONForms to generate the form builder, but halfway through we decided to hand-roll our own form builder. However, after experimentation in #188 and beyond, I come to realise that it is much more difficult to do our own implementation versus using JSONForms. Hence, the main trade-off being considered is:

> **Does the flexibility that comes with building our own form builder outweigh the effort needed to build and maintain it?**

| Building our own form builder | Using JSONForms |
|-----------|-------|
| **Pro**: Since we are building it ourselves, we can build it exactly the way we need it to fulfil the design requirements                                           | **Con**: We will need to understand the library, how it works and structure the code to produce the output we want |
| **Con**: A lot of the work involves parsing the JSON Schema in a manner that is extensible and safe, and it will be very difficult to do given the time constraints | **Pro**: JSONForms comes with a comprehensive API on working with the underlying JSON Schema |

**Examples:**

- In #188, the plan was to map through all the properties for the given component, then generate the appropriate form field based on the `type` and `format` keys for that prop. However, the approach of using `Object.keys` meant that we lose the types (since it is typed to `string`) when trying to work with each prop. For basic types, this is fine, but things can get quite messy when it comes to difficult ones (like those that involve `oneOf` and `$ref`).
- Speaking of `$ref`, it is a JSON Schema feature that allows us to reference another part of the schema, just like variables. Building the form builder ourselves means we will need to implement the functionality to resolve the schema, and I found that we are basically just copying [what JSONForms has already done](https://github.com/eclipsesource/jsonforms/blob/c00b664ced7bb17ca2ab838ce5da610790c76566/packages/core/src/util/resolvers.ts#L112).
- When it comes to the complicated types like `oneOf`, we will also need to implement our own handlers both on parsing the schema and handling the form state. Form state can be somewhat offloaded to react-hook-forms, but parsing the schema is still something we have to do (and again is something that [JSONForms has already done before](https://github.com/eclipsesource/jsonforms/blob/c00b664ced7bb17ca2ab838ce5da610790c76566/packages/material-renderers/src/complex/MaterialOneOfRenderer.tsx#L72-L79)).

**But why do we decide to hand-roll in the first place?**

- Mainly because JSONForms is a library that none of us have extensive experience with. Most of the work with the library was done by Jacob previously, and I only briefly touched it when working on Isomer Playground.
- The documentation for JSONForms is quite bad, the only way to know how to do things is to reference their repository code.
- We lose some flexibility if we use JSONForms, since it is a generic library that is intended to be used by all kinds of JSON Schemas.

**What made me decide to go back to JSONForms?**

- Mainly when trying to implement a function to handle `$ref`. I found that there is no other way than to just copy how JSONForms implemented `resolveSchema`, and we will need to duplicate the [types that they have already defined](https://github.com/eclipsesource/jsonforms/blob/c00b664ced7bb17ca2ab838ce5da610790c76566/packages/core/src/models/jsonSchema7.ts) in their library.
- I had some confidence in [their custom renderers](https://jsonforms.io/docs/tutorial/custom-renderers), having done something similar in Isomer Playground to get a Tiptap editor to work. Most of the time is just figuring out how to scope their testers and working with the available props from their API. Both are quite doable if we reference their repository code directly (which is what I did in this PR, referencing both their material and vanilla renderers).
- To put in perspective, I was able to get the basic types (#227) and placeholders for the more advanced types (#230) done in a single day. I spent longer trying to figure out how to do the same in our own hand-rolled version (#188) and some more time trying to figure out how to do things with type-safety.

Overall, I think that we should offload the handling of JSON Schema to JSONForms, so we reduce the overall maintenance burden and complexity in the longer term.

### CODE REVIEW

**Review by seaerchin — APPROVED** [review]
no issues on my end - actually this PR looks pretty clean! i'm ok w/ JSONForms for simple blocks (not requiring much validation); if it composes well, it should be seamless for complex blocks using `$ref`?

**Review by kishore03109 — APPROVED** [review]
I think your benefits for json forms is more than what you mentioned in your PR
JSON Forms have i18n enabled out of the box, allows us to use it in the future
They also have a sane definition of JsonSchema data type. I actually think this is non-trivial, since this would be hard to change down the line (we have been ping ponging on type definitions with this new code base). If we use their type def wholesale, we spend less time thinking about whether this is a sane type def.
I also think if we need to make changes && their renderer is not flexible enough, we can always fork out from json forms. Their code base is open sourced, so this is quite possible.
I also think as a team we need to be clear on product priorities! Ensuring type safety does make sense most of the time and should be on by default, but when we are dealing with dynamic input, it makes sense why type safety can be a PITA, but this should not block velocity. Note that JsonForms also have a bunch of ’any’s [esp when getting the shape](https://github.com/eclipsesource/jsonforms/blob/master/packages/core/src/generators/schema.ts#L29). Tldr; I think you made the right call here to switch back, maybe we should have documented why we made the original switch in the first place in PR/notion doc somewhere.

File: apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsIntegerControl.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+} from '@jsonforms/core'
+import { withJsonFormsControlProps } from '@jsonforms/react'
+import {
+  FormErrorMessage,
+  FormLabel,
+  NumberInput,
+} from '@opengovsg/design-system-react'
+
+export const jsonFormsIntegerControlTester: RankedTester = rankWith(
+  4,
```
[the remark this answers is not in the document]
Comment (dcshzj) [reply apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsIntegerControl.tsx:27]: Good point, I will tackle this in #230.

File: apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsIntegerControl.tsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  and(
+    uiTypeIs('Control'),
+    or(schemaTypeIs('integer'), schemaTypeIs('number')),
+    schemaMatches(
+      (schema) =>
+        (Object.prototype.hasOwnProperty.call(schema, 'maximum') ||
```
[the remark this answers is not in the document]
Comment (dcshzj) [reply apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsIntegerControl.tsx:33]: Clarified offline, answer is not sure cos I was referencing upstream library, will explore to see if it is safe for us to just do `schema.exclusiveMaximum`

**Pull request merged**